test/run_test.py

```diff
@@ -0,0 +1,228 @@
+#!/usr/bin/env python
+
+import argparse
+import os
+import shutil
+import subprocess
+import sys
+import tempfile
+
+import torch
+
+TESTS = [
+    'autograd',
+    'cpp_extensions',
+    'cuda',
+    'dataloader',
+    'distributed',
+    'distributions',
+    'indexing',
+    'jit',
+    'legacy_nn',
+    'multiprocessing',
+    'nccl',
+    'nn',
+    'optim',
+    'sparse',
+    'torch',
+    'utils',
+]
+
+WINDOWS_BLACKLIST = [
+    'cpp_extensions',
+    'distributed',
+]
+
+DISTRIBUTED_TESTS_CONFIG = {
+    'tcp': {
+        'WORLD_SIZE': '3'
+    },
+    'gloo': {
+        'WORLD_SIZE': '2' if torch.cuda.device_count() == 2 else '3'
+    },
+    'nccl': {
+        'WORLD_SIZE': '2'
+    },
+    'mpi': {},
+}
+
+
+def shell(command, cwd):
+    popen = subprocess.Popen(
+        command,
+        stdout=subprocess.PIPE,
+        universal_newlines=True,
+        cwd=cwd,
+        shell=True)
+    for stdout_line in iter(popen.stdout.readline, ''):
+        print(stdout_line.strip('\n'))
+    popen.stdout.close()
+    return_code = popen.wait()
+    return return_code == 0
+
+
+def get_shell_output(command):
+    return subprocess.check_output(command, shell=True).decode().strip()
+
+
+def run_test(python, test_module, test_directory, verbose):
+    verbose = '--verbose' if verbose else ''
+    shell('{} {} {}'.format(python, test_module, verbose), test_directory)
+
+
+def test_cpp_extensions(python, test_module, test_directory, verbose):
+    shell('{} setup.py install --root ./install'.format(python),
+          os.path.join(test_directory, 'cpp_extensions'))
+
+    python_path = os.environ.get('PYTHONPATH', '')
+    install_directory = get_shell_output(
+        "find cpp_extensions/install -name '*-packages'")
+    install_directory = os.path.join(test_directory, install_directory)
+    os.environ['PYTHONPATH'] = '{}:{}'.format(install_directory, python_path)
+    run_test(python, test_module, test_directory, verbose)
+    os.environ['PYTHONPATH'] = python_path
+
+
+def test_distributed(python, test_module, test_directory, verbose):
+    mpi_available = subprocess.call('command -v mpiexec', shell=True) == 0
+    if verbose and not mpi_available:
+        print('MPI not available -- MPI backend tests will be skipped')
+    for backend, env_vars in DISTRIBUTED_TESTS_CONFIG.items():
+        if backend == 'mpi' and not mpi_available:
+            continue
+        for with_init in {True, False}:
```

Comment:
nit: would be nicer to call it `with_file_init` (there's always *some* init, so it's unclear)